flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/DefaultLookupCache.java

```diff
@@ -0,0 +1,232 @@
+import org.apache.flink.shaded.guava30.com.google.common.cache.Cache;
+import org.apache.flink.shaded.guava30.com.google.common.cache.CacheBuilder;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.Collection;
+
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.CACHE_TYPE;
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.LookupCacheType.PARTIAL;
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.PARTIAL_CACHE_CACHE_MISSING_KEY;
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.PARTIAL_CACHE_EXPIRE_AFTER_ACCESS;
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.PARTIAL_CACHE_EXPIRE_AFTER_WRITE;
+import static org.apache.flink.table.connector.source.lookup.LookupOptions.PARTIAL_CACHE_MAX_ROWS;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Default implementation of {@link LookupCache}. */
+@PublicEvolving
+public class DefaultLookupCache implements LookupCache {
+    private static final long serialVersionUID = 6839408984141411172L;
+
+    // Configurations of the cache
+    private final Duration expireAfterAccessDuration;
+    private final Duration expireAfterWriteDuration;
+    private final Long maximumSize;
+    private final boolean cacheMissingKey;
+
+    // The underlying Guava cache implementation
+    private transient Cache<RowData, Collection<RowData>> guavaCache;
+
+    // Guava Ticker for testing expiration
+    private Ticker ticker;
+
+    // For tracking cache metrics
+    private Counter hitCounter;
+    private Counter missCounter;
+
+    private DefaultLookupCache(
+            Duration expireAfterAccessDuration,
+            Duration expireAfterWriteDuration,
+            Long maximumSize,
+            boolean cacheMissingKey) {
+        this.expireAfterAccessDuration = expireAfterAccessDuration;
+        this.expireAfterWriteDuration = expireAfterWriteDuration;
+        this.maximumSize = maximumSize;
+        this.cacheMissingKey = cacheMissingKey;
+    }
+
+    /** Creates a builder for the cache. */
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static DefaultLookupCache fromConfig(ReadableConfig config) {
+        checkArgument(
+                config.get(CACHE_TYPE).equals(PARTIAL),
+                "'%s' should be '%s' in order to build a default lookup cache",
+                CACHE_TYPE.key(),
+                PARTIAL);
+        return new DefaultLookupCache(
+                config.get(PARTIAL_CACHE_EXPIRE_AFTER_ACCESS),
+                config.get(PARTIAL_CACHE_EXPIRE_AFTER_WRITE),
+                config.get(PARTIAL_CACHE_MAX_ROWS),
+                config.get(PARTIAL_CACHE_CACHE_MISSING_KEY));
+    }
+
+    @Override
+    public void open(CacheMetricGroup metricGroup) {
```

Comment:
The `open()` method can be called concurrently and multiple times. So the implementation should be thread-safe, and we shouldn't initialize the cache again. Please add concurrent tests for `open()` and other methods.